### ISSUE TITLE
considering cherry-picking {{link-to}} component integration tests fix into 1.12

### ISSUE THREAD

**Issue by blimmer** [emberjs/ember.js#11825]
As originally noted in [ember-test-helpers#41](https://github.com/switchfly/ember-test-helpers/issues/41) there was an issue in Ember 1.12 with using the component integration tests on any component with a `{{link-to}}`. This issue was fixed by [emberjs/#11639](https://github.com/emberjs/ember.js/pull/11639) and released in the Ember 1.13.x cycle.

However, many of us are still making the transition Ember 1.12.x -> Ember 1.13.x because many add-ons are still implementing the changes necessary for Glimmer compatibility (specifically, our shim for ember-easy-forms doesn't work with glimmer and they're [not going to rewrite until Ember 2.0+](https://github.com/dockyard/ember-easy-form/issues/205)). Without the changes provided by [emberjs/#11639](https://github.com/emberjs/ember.js/pull/11639) in Ember 1.12, it limits our ability to stay up-to-date with ember-test-helpers and to take advantage of the (extremely) helpful component integration tests, which are now generated by default in ember-mocha and ember-qunit.

Is it possible to cherry-pick the aforementioned [PR](https://github.com/emberjs/ember.js/pull/11639) into 1.12 and issue a 1.12.2?

I realize that this is outside the realm of what we'd normally do in the Ember ecosystem, but I think because 1.12.x -> 1.13.x was a large one for many in the community, it warrants consideration. Thanks for the consideration.

/cc @rwjblue

**Comment by rwjblue** [comment]
Thanks for bringing this up, I'll ping the rest of the team to see what their thoughts are.

**Comment by stefanpenner** [comment]
@rwjblue we should likely branch 1.12 and 1.13 (once 2.0 becomes stable), so its easier for community backports etc.

**Comment by rwjblue** [comment]
Yes, my plan in general is to branch when we need to maintain a tag after tagging.  We really haven't had to do it yet, so that might not be obvious from the repo branches.

**Comment by blimmer** [comment]
Dumb question: is the v1.12.1 tag different from what a 1.12 branch (as Stef mentioned) would be?

**Comment by rwjblue** [comment]
@blimmer - Please submit a PR with those commits against the `stable-1-12` branch.

**Comment by blimmer** [comment]
Now that I tried to do the cherry-pick, I'm realizing that these commits will not cherry-pick cleanly because of the [routing service introduced with 1.13](https://github.com/nathanhammond/ember.js/commit/7cff945c04bf66b0c621b6b70042c51d8a5b2cdf). Unfortunately, I don't know the guts of ember very well, but I can try to see if there's a clear way forward.

**Comment by blimmer** [comment]
For others following along, here's a [simple example](https://github.com/blimmer/ember-cli-example-app-for-github/tree/component-link-to-integration-test) of where it's broken and the stacktrace:
```
TypeError: Cannot read property 'recognizer' of undefined
    at computeLinkViewRouteArgs (http://localhost:7357/assets/vendor.js:29632:204)
    at Descriptor.ComputedPropertyPrototype.get (http://localhost:7357/assets/vendor.js:20511:26)
    at Object.get (http://localhost:7357/assets/vendor.js:25730:19)
    at computeLinkViewLoading (http://localhost:7357/assets/vendor.js:29463:25)
    at Descriptor.ComputedPropertyPrototype.get (http://localhost:7357/assets/vendor.js:20511:26)
    at Object.get (http://localhost:7357/assets/vendor.js:25730:19)
    at computeActive (http://localhost:7357/assets/vendor.js:29715:22)
    at computeLinkViewActive (http://localhost:7357/assets/vendor.js:29419:14)
    at Descriptor.ComputedPropertyPrototype.get (http://localhost:7357/assets/vendor.js:20511:26)
    at Object.get (http://localhost:7357/assets/vendor.js:25730:19)
```

**Comment by blimmer** [comment]
So it's not a cherry-picked fix, but I got this working. See #11829

**Comment by tomdale** [comment]
@blimmer Doing a point release for 1.12 is highly unusual and I'd prefer not to set that precedent for anything but security releases. What is keeping you on 1.12? I'd rather prioritize fixing issues in 1.13 that are preventing you from upgrading.

**Comment by blimmer** [comment]
I bet on [ember-easy-form](https://github.com/dockyard/ember-easy-form) about a year ago and it has been the single biggest point of pain in upgrading Ember versions since handlebars 2.0 and Ember 1.8. I've been hobbling along using [a shim](https://github.com/dockyard/ember-easy-form/issues/175#issuecomment-76475137) that I've hacked version-by-version until now.

In order to get off of easy-forms I'd need to rewrite every form in my production app, which would take about a week and I can't sell that to my boss. It seems like they are blocked by [scoped helpers](https://github.com/emberjs/ember.js/pull/10244) and they [won't fix the addon](https://github.com/dockyard/ember-easy-form/issues/205#issuecomment-114376792) until they get them.

ef4 showed me how to publish [a fork](https://github.com/Ibotta/ember) which I'm using right now in my app.

**Comment by stefanpenner** [comment]
@bcardarella out of curiosity would you accept fixes/patches/adjustments to simple-form, so @blimmer and likely many others don't stay stuck?

**Comment by bcardarella** [comment]
@stefanpenner the easy-form issues stem around 1.10 - 1.13 and the move to HTMLBars. Specifically, there could be a interim release to help ease migrations that packaged compiled HTMLBars templates. If someone wants to do that I can merge and release.

As far as ember-easy-form 2.0 and ember-cli I am not willing to move forward until scoped helpers are in Ember https://github.com/emberjs/ember.js/pull/10244

**Comment by tsing80** [comment]
+1 for merging it into 1.12 too. There are still addons (such as ember-table)  that is not ported to 1.13 yet which block me upgrading to 1.13

**Comment by stefanpenner** [comment]
> +1 for merging it into 1.12 too.

this is unlikely to happen

> which block me upgrading to 1.13

merging this is a similar fix into 1.13 is a stretch but may be possible

**Comment by rwjblue** [comment]
@stefanpenner - This is already fixed in 1.13.

**Comment by stefanpenner** [comment]
> @stefanpenner - This is already fixed in 1.13.

perfect.

**Comment by mixonic** [comment]
I've opened a fix for this in ember-test-helpers: https://github.com/switchfly/ember-test-helpers/pull/100

**Comment by blimmer** [comment]
Thank you!!

**Comment by blimmer** [comment]
I upgraded ember test helpers and that doesn't resolve this problem. I think the use of the fork is necessary still.

**Comment by rwjblue** [comment]
A test was added in switchfly/ember-test-helpers#100 confirming that `{{link-to}}` can be rendered in a component integration test, and that test passes on 1.10, 1.11, 1.12, 1.13, 2.0, 2.1-betas, and canary.

The test does:

```
test('it can render a link-to', function() {
  this.render("{{link-to 'Hi' 'index'}}");
  ok(true, 'it renders without fail');
});
```

@blimmer - Can you add a few more tests over there, that show what is failing?

**Comment by blimmer** [comment]
I spent some time trying to figure out why the written test isn't failing, and I couldn't figure it out. I did, however, make a very simple project that shows it's still failing with the most current ember-qunit, which should be using the more current ember-test-helpers.

Maybe I'm missing something, but it wasn't obvious after looking for about 30 minutes.

[Failing tests](https://travis-ci.org/blimmer/ember-cli-example-app-for-github/builds/80761509)
[Branch of code](https://github.com/blimmer/ember-cli-example-app-for-github/commit/4e30f876bf62beb6719a84550996f70fb3d753d6)

Easy repro steps:
1. `ember new my-app`
2. change bower.json to ember 1.12.1 and ember-qunit 0.4.11 (should have ~0.5.10 ember-test-helpers)
3. `ember g route about`
4. `ember g component foo-bar`
5. change `foo-bar.hbs` to this: ```{{#link-to 'about}}About{{/link-to}}````
6. `ember t`

Still fails with this message:
```
at http://localhost:7357/assets/test-support.js:6499: 'undefined' is not an object (evaluating 'router.router.recognizer')
```
